lib/queuePopulator/IngestionProducer.js

```diff
@@ -185,95 +256,142 @@ class IngestionProducer {
     }
 
     /**
-     * get the list of objects for each bucket
+     * Get the list of object versions for a bucket
      *
-     * @param {object} bucketList - list of buckets
+     * @param {string} bucket - bucket name
+     * @param {object} state - previous state used to paginate version listing
+     * @param {string} [state.versionMarker] - NextVersionIdMarker
+     * @param {string} [state.keyMarker] - NextKeyMarker
      * @param {function} done - callback function
-     * @return {object} list of buckets and list of objects for each bucket
+     * @return {object} list of objects for each bucket, including a duplicate
+     *   entry for IsLatest versions
      */
-    _getBucketObjects(bucketList, done) {
-        if (!bucketList) {
-            return done(null, null);
+    _getObjectVersionsList(bucket, state, done) {
+        if (bucket === constants.usersBucket ||
+            bucket === constants.metastoreBucket) {
+            return done();
         }
-        return async.mapLimit(bucketList, 10, (bucketInfo, cb) => {
-            if (bucketInfo === constants.usersBucket ||
-            bucketInfo === constants.metastoreBucket) {
-                return cb();
+        const { versionMarker, keyMarker } = state;
+        const params = {
+            Bucket: bucket,
+        };
+        // if previous state, should paginate here
+        if (versionMarker && keyMarker) {
+            params.VersionIdMarker = versionMarker;
```

Comment:
What happens if the version had been deleted or the objects in the bucket were modified. Is it okay to not have the listing continue from where it had left off before?

Reply:
For modifications, should be fine. This initial listing/snapshot phase is a best attempt at capturing objects quickly. Then the md logs should pick up the rest of the objects. Duplicating entries is also fine since we plug right into mongo.

Good point on deleted version. I just checked and it seems if the marker is deleted (not a delete marker), the marker will still work and list from the expected spot. I'm guessing the version id's are in order. If the version id marker is provided but doesn't exist, it will know to just list by whichever is next in the listing

Reply:
For duplicates, if you overwrite the metadata of an object that had been replicated by Zenko, would you end up putting the object twice? Does backbeat log processor handle that scenario?

Reply:
ah, good call, the log processor won't handle this scenario. There is a plan to add a `headObject` check on processor side (for the PR applying replication info on each object md before saving to mongo).

I'm thinking, we will need to get object md from mongo instead of doing the `headObject` call, and seeing if replication info has already been set. If so, keep the status. But i'll need to update other md only fields in-case this is an md update from s3c.

Reply:
I agree there would need to be some logic around overwriting any existing metadata. 

I've been working intermittently on a metadata locking mechanism that might be applicable for this purpose. It allows you to update metadata only on certain conditions and handles race conditions when multiple clients are attempting concurrent writes to the same metadata.

Reply:
nice! that sounds useful for sure